src/python/pants/backend/project_info/tasks/export_dep_as_jar.py

```diff
@@ -0,0 +1,301 @@
+      'platforms': {
+        str(platform_name): {
+          'target_level': str(platform.target_level),
+          'source_level': str(platform.source_level),
+          'args': platform.args,
+        } for platform_name, platform in JvmPlatform.global_instance().platforms_by_name.items()},
+    }
+
+    graph_info = {
+      'version': DEFAULT_EXPORT_VERSION,
+      'targets': targets_map,
+      'jvm_platforms': jvm_platforms_map,
+      'scala_platform': scala_platform_map,
+      # `jvm_distributions` are static distribution settings from config,
+      # `preferred_jvm_distributions` are distributions that pants actually uses for the
+      # given platform setting.
+      'preferred_jvm_distributions': {}
+    }
+
+    for platform_name, platform in JvmPlatform.global_instance().platforms_by_name.items():
+      preferred_distributions = {}
+      for strict, strict_key in [(True, 'strict'), (False, 'non_strict')]:
+        try:
+          dist = JvmPlatform.preferred_jvm_distribution([platform], strict=strict)
+          preferred_distributions[strict_key] = dist.home
+        except DistributionLocator.Error:
+          pass
+
+      if preferred_distributions:
+        graph_info['preferred_jvm_distributions'][platform_name] = preferred_distributions
+
+    if runtime_classpath:
+      graph_info['libraries'] = self._resolve_jars_info(targets, runtime_classpath)
+      # Using resolved path in preparation for VCFS.
+      resource_jar_root = os.path.realpath(self.versioned_workdir)
+      for t in targets:
+        target_type = _get_target_type(t)
+        # If it is a target root or it is already a jar_library target, then no-op.
+        if t in target_roots_set or targets_map[t.address.spec]['pants_target_type'] == 'jar_library':
+          continue
+
+        if target_type == SourceRootTypes.RESOURCE or target_type == SourceRootTypes.TEST_RESOURCE:
+          # yic assumed that the cost to fingerprint the target may not be that lower than
+          # just zipping up the resources anyway.
+          with temporary_file(root_dir=resource_jar_root, cleanup=False, suffix='.jar') as f:
+            with zipfile.ZipFile(f, 'a') as zip_file:
+              for src_from_source_root, src_from_build_root in zip(t.sources_relative_to_source_root(), t.sources_relative_to_buildroot()):
+                zip_file.write(os.path.join(get_buildroot(), src_from_build_root), src_from_source_root)
+          targets_map[t.address.spec]['pants_target_type'] = 'jar_library'
+          targets_map[t.address.spec]['libraries'] = [t.id]
+          graph_info['libraries'][t.id]['default'] = f.name
+          continue
+
+        targets_map[t.address.spec]['pants_target_type'] = 'jar_library'
+        targets_map[t.address.spec]['libraries'] = [t.id]
+        jar_products = runtime_classpath.get_for_target(t)
+        for conf, jar_entry in jar_products:
+          if 'z.jar' in jar_entry:
```

Comment:
Is this jar always produced when we compile any target, or is there a flag that's necessary?

Reply:
added some comments and TODO.